### PR TITLE
Allow loading gzipped data files

### PR DESCRIPTION
**Pull request by mike-seger**
## Impact
* [ ]  Bug fix (non-breaking change which fixes expected existing functionality)
* [x]  Enhancement/New feature (adds functionality without impacting existing logic)
* [ ]  Breaking change (fix or feature that would cause existing functionality to change)

## Description
Allows loading gzipped data files.
The change is similar to https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/resource/PathResource.java#L34

For example, when using:
```
databaseChangeLog:
  - changeSet:
      id: load data
      author: user1
      changes:
        - loadData:
            file: ./data/CITY.csv.gz
```

the extension ".gz" switches reading the content as gzipped data.

## Things to be aware of
* Should an application be using  data files with names ending in ".gz" which are **NOT** gzipped, the file name has to be changed to a proper extension.

## Things to worry about
* Nothing

### CODE REVIEW

**Comment by nvoxland** [comment]
**Pre-Review Notes**

Change makes sense. Probably good to add a loadData with a .gz file to the `common.tests.changelog.xml` file to test out the e2e flow through it

Questions I have:
- I thought there was maybe gz support before, did that get broken? Or was I just mis-remembering?

Potential risks:
- None

What could make the full review difficult:
- Nothing

**Comment by mike-seger** [comment]
@nvoxland: 
You're right that similar functionality exists in liquibase in at least one place, where an entire directory can be marked as containing gzipped resources. There are several other places and ways liquibase loads resources, which imho require a re-factoring to consolidate them into a single place.   
In order to remain low impact, this change would improve the situation a lot for the time being.

**Comment by nvoxland** [comment]
@mike-seger I agree. Nice to keep this more isolated and consider a larger cleanup later.

**Comment by filipelautert** [comment]
Hello @mike-seger ! Thanks for the code, I also added the functional tests to the branch.

But when I added the tests, I noticed that the "resource" being used is of type PathResource that already handles gzip files as from the description link https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/resource/PathResource.java#L34 .

The fix made it into liquibase 4.17.0 , did you try it with this version? As i see your PR is from around the days it was released.

**Comment by mike-seger** [comment]
Hello @filipelautert , thanks for adding an integration test.  
I checked whether liquibase 4.17.0 would actually resolve this case. Here's what happens:
```
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset 20220803-210000.yaml::load data::adminux:

     Reason: liquibase.exception.UnexpectedLiquibaseException: CSV file ./data/CITY.csv.gz Line 2 has 1 values defined, Header has 3. Numbers MUST be equal (check for unquoted string with embedded commas)
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:126) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Liquibase.lambda$null$0(Liquibase.java:291) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Liquibase.lambda$update$1(Liquibase.java:290) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Liquibase.runInScope(Liquibase.java:2447) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Liquibase.update(Liquibase.java:236) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Liquibase.update(Liquibase.java:221) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:328) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:283) ~[liquibase-core-4.17.0.jar:na]
	at org.springframework.boot.autoconfigure.liquibase.DataSourceClosingSpringLiquibase.afterPropertiesSet(DataSourceClosingSpringLiquibase.java:46) ~[spring-boot-autoconfigure-2.7.4.jar:2.7.4]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863) ~[spring-beans-5.3.23.jar:5.3.23]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800) ~[spring-beans-5.3.23.jar:5.3.23]
	... 18 common frames omitted
Caused by: liquibase.exception.MigrationFailedException: Migration failed for changeset 20220803-210000.yaml::load data::adminux:
     Reason: liquibase.exception.UnexpectedLiquibaseException: CSV file ./data/CITY.csv.gz Line 2 has 1 values defined, Header has 3. Numbers MUST be equal (check for unquoted string with embedded commas)
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:714) ~[liquibase-core-4.17.0.jar:na]
Caused by: liquibase.exception.MigrationFailedException: Migration failed for changeset 20220803-210000.yaml::load data::adminux:

	at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:56) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.changelog.ChangeLogIterator$2.lambda$null$0(ChangeLogIterator.java:113) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.changelog.ChangeLogIterator$2.lambda$run$1(ChangeLogIterator.java:112) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:93) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.lambda$child$0(Scope.java:180) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:189) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:179) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:158) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:243) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.Scope.child(Scope.java:247) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:65) ~[liquibase-core-4.17.0.jar:na]
	... 37 common frames omitted
Caused by: liquibase.exception.UnexpectedLiquibaseException: CSV file ./data/CITY.csv.gz Line 2 has 1 values defined, Header has 3. Numbers MUST be equal (check for unquoted string with embedded commas)
Caused by: liquibase.exception.UnexpectedLiquibaseException: CSV file ./data/CITY.csv.gz Line 2 has 1 values defined, Header has 3. Numbers MUST be equal (check for unquoted string with embedded commas)

	at liquibase.change.core.LoadDataChange.generateStatements(LoadDataChange.java:311) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1250) ~[liquibase-core-4.17.0.jar:na]
	at liquibase.changelog.ChangeSet.execute(ChangeSet.java:679) ~[liquibase-core-4.17.0.jar:na]
	... 57 common frames omitted
```

This happens because:
liquibase.change.core.LoadDataChange.getCSVReader(). 
calls:
ResourceAccessor resourceAccessor = Scope.getCurrentScope().getResourceAccessor();  
which does not have any gzip decoding.

Maybe it would make sense to add another bug/cleanup request to consolidate all inputstream handling in one place. It could even be smart and not even rely on the file name, but detect the type of stream (such as gzip) by inspecting the first few bytes using PushbackInputStream.

As this is possibly a major refactoring endeavour, I'm not expecting this to happen soon.

Bottom line, it would be great if this PR could make it into liquibase for the time being.

**Comment by filipelautert** [comment]
Hey @mike-seger - I just added another check to make sure it works in other cases, so it may still work for you now. If you think it's fine I'll move it ahead for further code review and integration. Thanks!

**Comment by mike-seger** [comment]
@filipelautert The change is still looking good with me

**Review by filipelautert — APPROVED** [review]
* Change makes sure that we don't break what we currently have;
*  Add support to gzipped input stream
* Tests added.

**Review by suryaaki2 — APPROVED** [review]

**Review by XDelphiGrl — APPROVED** [review]
Want to `loadData` from a gzip archive? This PR makes that possible by extending the `LoadDataChange.getCSVReader()` method to check for `.gz` files and inflate the archive. There is a comment from @mike-seger that there is a less-brittle way than using the file extension to determine that the data is coming from a gzip archive; I'm going to approve this fix as-is because the tests are passing and it unblocks people who would like to use this new feature.
* Automated functional and test harness executions passing.
* New integration test added to test `loadData` with data coming from a `.gz` archive.
* No additional testing required.

@filipelautert and @MalloD12, if you think it makes sense at this time to update the code to use Mike's suggestion, please do. If so, ping me when the rework is done. I'm not formally requesting changes to the PR.

APPROVED
